### PR TITLE
"ESP32 Firmware Settings" tab

### PR DESCRIPTION
**Pull request by KKawase0104**
I have found that it is difficult for beginner to use the command prompt for update MicroPython to ESP32.

I Added "ESP32 Firmware Settings" tab to Mu Administration dialog.
The user selects a MicroPython file by "..." button and upload MicroPython to ESP32 by "update" button. If "Erase the entire flash before updating?" checkbox is on, erase the entire flash before update firmware.

image

The erase and upload command are same below.
https://micropython.org/download#esp32

As this function use esptool.py, the user must install esptool by "Third Party Packages" tab.

### CODE REVIEW

**Comment by dybber** [comment]
Great! Teachers really need this, to make ESP MicroPython easier to adopt. I hope to find time to test it soon. One initial comment: Would it make sense if the tab was only shown if the user has selected the ESP MicroPython-mode?

PS. I'm currently working on making auto-detecting of devices smoother with a menu for selecting the current device.

**Comment by KKawase0104** [comment]
Thank you for your comment and help!
Thanks to your advice, I have thought it is useful to show this tab only in ESP MicroPython-mode.
I hope to find time to add its function.
Thank you.

**Comment by dybber** [comment]
Hi @KKawase0104 

I finally got around to testing your pull request, and it seems to work very fine for me and a Sparkfun ESP32 Thing. However, it didn't work on an ESP8266-device. It seems that you are making esptool do an auto-detection for the erase_flash part, but for the write_flash you very specifically mention `--chip esp32`.

In addition, I think a key thing is also to make the user-interface self-explanatory, such that most teachers can figure out to use it without too much trouble. So my main feedback is about the user-interface:

 - The other admin panels, begins with a description. I think this one should also have a very brief description
 - I think the text on how to check the current version and install esptool, should also be above the panel, the panel should only be for esptool-output in my opinion, using it for anything else is just confusing.
 - Perhaps it could automatically detect whether esptool is already installed, and present a warning and disable the Update-button as long as it is still not installed?
 - I think the "Update" button should have a different label, for example "Flash firmware"
 - When we get it to work on other device (ESP8266) i think the name should probably not mention ESP32, perhaps just "Firmware flasher".
 - Finally, it prints the following when you flash, which I guess it shouldn't:
```('/Users/dpr964/Downloads/esp32-idf3-20191220-v1.12.bin', 'Firmware (*.bin)')
['/Users/dpr964/Library/Application Support/mu/site-packages/esptool.py', 'erase_flash']
['/Users/dpr964/Library/Application Support/mu/site-packages/esptool.py', '--chip', 'esp32', '--baud', '460800', 'write_flash', '-z', '0x1000', '/Users/dpr964/Downloads/esp32-idf3-20191220-v1.12.bin']```

I can also help work on this, if you want help, I have some free time over the next couple of weeks.

Perhaps when we get this into a more final state, we could also write a tutorial for codewith.mu? Then this admin panel could link there for a more in-depth description.

**Comment by KKawase0104** [comment]
Hi @dybber 

Thank you for your help and advice!
Since it is first time for me to do this contribution, your advice is very helpful.
I am so glad you even offered help for me!
I want help!

Thank you again!

**Comment by dybber** [comment]
Hi @KKawase0104 
I had a look at it today, and have tried moving a bit around, adding instructions and I even found a way to check whether the `esptool` is installed programmatically.

The branch with my suggested changes are in: https://github.com/dybber/mu/tree/feature/esp_firmware_flashing

Users are now required to also choose whether it's an ESP8266 or ESP32 device, on the otherhand I've chosen to do simplification regarding erase_flash. Instead of allowing users to choose, it just always does both an erase_flash and a write_flash. I guess this will work for almost any user, power-users can still use esptool.py directly.

Below are two screenshots. Let me know if you have any comments. I still need to make some tests with this, but I'm pretty happy with our combined efforts!

Best regards,
Martin

Screenshot:
Screenshot 2020-03-26 at 23 32 51

If `esptool` is not installed they will instead see:
Screenshot 2020-03-27 at 00 05 42

**Comment by KKawase0104** [comment]
Hi Martin (@dybber)

User-friendly GUI !! 
Your source code will be very learning for me.
I'm pretty happy, too.

I had run your ESP Firmware flasher.
As 'exit' value of 'esptool_finished' method is undefined, flash process will stop after erase.
I know you are test phase. Sorry.

Best regards,
Kenji

**Comment by dybber** [comment]
Thanks for testing, I've just pushed a fix. It was a typo, should've written "exitCode" instead of just "exit". Let me know if you experience any other problems.

**Comment by KKawase0104** [comment]
OK. I don't know why, but in my environment, the second screenshot doesn't show up when esptool isn't installed. That's all I have so far.

**Comment by dybber** [comment]
Which platform are you on? I'm on Mac OS X, maybe there is some difference there.

**Comment by KKawase0104** [comment]
I use Windows10. I hope to find time to test in my environment.

**Comment by KKawase0104** [comment]
Hello.
I looked into the problem.
The reason was that the pkguil.find_loader around line 247 of dialogs.py did not refer to Mu's Python.The 2nd screenshot came up by changing it below.
esptool_installed = os.path.exists(MODULE_DIR + "/esptool.py")

The Erase process worked correctly, but the firmware could not be written.
I will look at it again when I have time.

**Comment by KKawase0104** [comment]
Hello,
I'm very sorry. The Firmware writing was correct. I selected different Device type.

**Comment by the-stanely** [comment]
@dybber , thank you for inviting me to test this feature.  Sorry for the delay.  I tested this with two flavors of ESP32 dev boards, LILY GO TTGO (T-Display version), and ESP32 DevKitC.  They were new, in static bag devices with no uPython installed.  Firmware installation was flawless and fast.  After flashing, REPL worked just like expected.  Each ESP32 reported its firmware version on boot.  "Files" showed only boot.py in the device folder.

This made MicroPython installation really, really easy!

I played with re-installing MicroPython too.  Because uP was already installed, I left "Files" panel opened, and connect failed.  Same thing with leaving the REPL panel open.  An unable to connect message was shown and then it kept trying but I didn't wait for timeout.  I don't know if it's practical to check that before trying the erase/flash.

Then I closed the bottom panels and tried again but forgot to set ESP32.  It failed with this messages:
...
WARNING: Detected crystal freq 20.22MHz is quite different to normalized freq 26MHz. Unsupported crystal in use?
Crystal is 26MHz
MAC: 00:08:ff:00:00:00
Uploading stub...
Running stub...

A fatal error occurred: Invalid head of packet (0x46)

Error on flashing. Aborting.
...
So that's probably good.  I changed the processor to ESP32 and all went well.

There was one thing that was difficult to read.  In the two dark themes, the link to the micropython binaries is dark blue.  It is not legible on my monitor.  Don't know if anything needs to be done about that.

Please let me know if there is anything specific you want me to test.

**Comment by dybber** [comment]
@KKawase0104, thanks for that suggestion. That new way to detect presence of esptool.py, also seems to work on my Mac. I've just committed the change

@the-stanely Great find with the REPL/Files pane open, that should of course not be possible. For the users I guess, the simplest thing would just be to automatically close either REPL or Files pane when pushing erase flash? Or should there be a warning that they will be closed? My feeling would be that it would just be superfluous to ask, when the user have already pushed "Erase & Flash".

I also looked at the issue with the link color, unfortunately it's not possible to change link colors in the stylesheets we swap between when changing theme. However, I changed the link color of this specific link a little more light blue. That makes it a bit more difficult to see in the day theme, but not that much, as you can see here:
Screenshot 2020-04-13 at 21 40 42

**Comment by KKawase0104** [comment]
@dybber , Thank you, I've confirmed that it works.

@the-stanely , Thank you for time. I'm glad we developed it together.

I thought that REPL/Files/Plotter Pane should close automatically when Mu Administration button pushed. But is this way not good because Mu Administration is used all mode ?

**Comment by dybber** [comment]
I think that closing the REPL/Files/Plotter when administration button is pushed, wouldn't be ideal for everybody. The administration window also displays the log file, and if you are trying to understand what problems you are facing with the REPL, then it wouldn't be good if the administration button closed the connection to the device. Therefore I think it should happen at the last possible time - that is, when pushing "Erase & write firmware"

**Comment by KKawase0104** [comment]
Thank you for your reply. I see the reason about panel close timing. Thank you again.

**Comment by dybber** [comment]
Great! Your recent commit seems to do the job.

I guess the next thing is to update the test suite. To be merged into master, "make check" shouldn't report any errors and have full coverage.

**Comment by dybber** [comment]
Btw. you should probably also remove the debug prints ("print('repl')", etc.) from your code. If you want to log some debug/info/warnings/error messages, use the logger (logger.info() for instance)

**Comment by KKawase0104** [comment]
Thank you for you reply!
Very good intuition. 
I work educational material making company.
My company has developed programmable educational toy "ArtecRobo" which has ESP32.
https://www.artec-kk.co.jp/artecrobo2/en/
I have added our robot mode to Mu and sent PR.
https://github.com/mu-editor/mu/pull/922
So, I know inside of Mu a little.
It is happy for us to make Mu better by learning.

I am very happy to see you and @the-stanely because you have a lot of thing I don't have. I'm learning.

I will do the next thing and remove the debug prints. (^^;;)

Thank you.

**Comment by dybber** [comment]
Great, I think with your latest commit this should be ready, at least I don't have anymore comments/suggestions. Looking forward to get this integrated.

PS. Sounds like a fun place to be working @KKawase0104!

**Comment by KKawase0104** [comment]
Thank you very much.
I'll try to add other features in the future.

File: mu/resources/css/day.css [hunk 1]
```diff
@@ -278,3 +278,27 @@ QTreeView::branch:has-children:open {
...
+QGroupBox {
+    border: 1px solid #b4b4b4;
+    margin-top: 1ex; /* leave space at the top for the title */
+}
+
+QGroupBox::title {
+    subcontrol-origin: margin;
+    left: 8px;
+}
+
```
[the remark this answers is not in the document]
Comment (KKawase0104) [reply mu/resources/css/day.css:304]: Thank you for your time.
I have reflected review on f236a83.

**Comment by dybber** [comment]
Hi @KKawase0104, I'm closing this as I've replaced it with a new version #1100, which I think we can merge very soon, when someone else have tested it and confirmed that it still works. See my edits over in #1100, and thank you for the contribution!

**Comment by KKawase0104** [comment]
@dybber OK. I'm happy to contribute to Mu. Thank you.

**Pull request closed without merging**